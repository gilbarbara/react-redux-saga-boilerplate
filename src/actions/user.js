// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { userLogin: login, userLogout: logOut } = createActions({
  [ActionTypes.USER_LOGIN]: () => ({}),
  [ActionTypes.USER_LOGOUT]: () => ({}),
});
