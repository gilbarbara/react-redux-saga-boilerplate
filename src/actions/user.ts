import { createAction } from '@reduxjs/toolkit';

import { ActionTypes } from '~/literals';

export const login = createAction(ActionTypes.USER_LOGIN_REQUEST);
export const loginSuccess = createAction(ActionTypes.USER_LOGIN_SUCCESS);

export const logOut = createAction(ActionTypes.USER_LOGOUT_REQUEST);
export const logOutSuccess = createAction(ActionTypes.USER_LOGOUT_SUCCESS);
