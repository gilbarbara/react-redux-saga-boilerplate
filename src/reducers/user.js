import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';

import { STATUS, ActionTypes } from 'constants/index';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGIN_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: true },
          status: { $set: STATUS.READY },
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: STATUS.RUNNING },
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          status: { $set: STATUS.IDLE },
        }),
    },
    userState,
  ),
};
