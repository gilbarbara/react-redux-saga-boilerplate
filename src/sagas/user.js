/**
 * @module Sagas/User
 * @desc User
 */

import { delay } from 'redux-saga';
import { all, call, put, takeLatest } from 'redux-saga/effects';

import { ActionTypes } from 'constants/index';

/**
 * Login
 */
export function* login() {
  try {
    yield call(delay, 400);

    yield put({
      type: ActionTypes.USER_LOGIN_SUCCESS,
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN_FAILURE,
      payload: err,
    });
  }
}

/**
 * Logout
 */
export function* logout() {
  try {
    yield call(delay, 200);

    yield put({
      type: ActionTypes.USER_LOGOUT_SUCCESS,
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGOUT_FAILURE,
      payload: err,
    });
  }
}

/**
 * User Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN, login),
    takeLatest(ActionTypes.USER_LOGOUT, logout),
  ]);
}
