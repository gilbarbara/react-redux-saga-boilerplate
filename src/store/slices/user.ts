import { createSlice } from '@reduxjs/toolkit';

import { STATUS } from '~/literals';

import { UserState } from '~/types';

export const userState: UserState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export const userSlice = createSlice({
  name: 'user',
  initialState: userState,
  reducers: {
    login: draft => {
      draft.status = STATUS.RUNNING;
    },
    loginSuccess: draft => {
      draft.isAuthenticated = true;
      draft.status = STATUS.READY;
    },
    logOut: draft => {
      draft.status = STATUS.RUNNING;
    },
    logOutSuccess: draft => {
      draft.isAuthenticated = false;
      draft.status = STATUS.IDLE;
    },
  },
});

export const { login, loginSuccess, logOut, logOutSuccess } = userSlice.actions;
export default userSlice.reducer;
