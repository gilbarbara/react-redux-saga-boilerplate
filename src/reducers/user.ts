import { createReducer } from '@reduxjs/toolkit';

import { login, loginSuccess, logOut, logOutSuccess } from '~/actions';
import { STATUS } from '~/literals';

import { UserState } from '~/types';

export const userState = {
  isAuthenticated: false,
  status: STATUS.IDLE,
};

export default {
  user: createReducer<UserState>(userState, builder => {
    builder
      .addCase(login, draft => {
        draft.status = STATUS.RUNNING;
      })
      .addCase(loginSuccess, draft => {
        draft.isAuthenticated = true;
        draft.status = STATUS.READY;
      });

    builder
      .addCase(logOut, draft => {
        draft.status = STATUS.RUNNING;
      })
      .addCase(logOutSuccess, draft => {
        draft.isAuthenticated = false;
        draft.status = STATUS.IDLE;
      });
  }),
};
