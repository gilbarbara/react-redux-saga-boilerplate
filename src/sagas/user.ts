import { all, delay, put, takeLatest } from 'redux-saga/effects';

import { loginSuccess, logOutSuccess } from '~/actions';
import { ActionTypes } from '~/literals';

export function* loginSaga() {
  yield delay(400);

  yield put(loginSuccess());
}

export function* logoutSaga() {
  yield delay(200);

  yield put(logOutSuccess());
}

export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN_REQUEST, loginSaga),
    takeLatest(ActionTypes.USER_LOGOUT_REQUEST, logoutSaga),
  ]);
}
