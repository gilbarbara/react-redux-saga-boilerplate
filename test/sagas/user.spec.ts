import { delay } from 'redux-saga/effects';
import { expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';

import user, { login, logout } from 'sagas/user';

describe('user', () => {
  it('should have the expected watchers', () =>
    expectSaga(user)
      .run({ silenceTimeout: true })
      .then(result => {
        expect(result.toJSON()).toMatchSnapshot();
      }));

  describe('login', () => {
    it('should handle SUCCESS', () =>
      expectSaga(login)
        .provide([[matchers.call.fn(delay), throwError(new Error('fail'))]])
        .run(400)
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        }));
  });

  describe('logout', () => {
    it('should handle SUCCESS', () =>
      expectSaga(logout)
        .run({ silenceTimeout: true })
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        }));
  });
});
