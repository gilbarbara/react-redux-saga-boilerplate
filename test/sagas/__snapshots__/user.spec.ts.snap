// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`user > login > should handle SUCCESS 1`] = `
{
  "call": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "args": [
          400,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
  ],
  "put": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "action": {
          "payload": undefined,
          "type": "USER_LOGIN_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`user > logout > should handle SUCCESS 1`] = `
{
  "call": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "args": [
          200,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/delayP",
      },
      "type": "CALL",
    },
  ],
  "put": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "action": {
          "payload": undefined,
          "type": "USER_LOGOUT_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`user > should have the expected watchers 1`] = `
{
  "fork": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "args": [
          "USER_LOGIN_REQUEST",
          "@@redux-saga-test-plan/json/function/loginSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "args": [
          "USER_LOGOUT_REQUEST",
          "@@redux-saga-test-plan/json/function/logoutSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": [
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "pattern": "USER_LOGIN_REQUEST",
      },
      "type": "TAKE",
    },
    {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": {
        "pattern": "USER_LOGOUT_REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;
